### ISSUE TITLE
Add per-epoch learning rate schedules

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#3503]
(I thought we already had an issue on this, but apparently not)

LR schedules: a common use case is as follows:
"After X epochs, set LR to A, after Y epochs set to B".
Currently this is possible - but only if we know the number of iterations per epoch and set it using that.

Note that we should log to console when the LR schedules are applied per epoch too...
"Applying LR schedule at start of epoch X; N layers impacted" or  something like that.

Edit: while we are at it, let's make LR schedules possible to apply according to any of the following:
- Iterations (parameter updates)
- Minibatches (same as iterations, *except* when doing TBPTT)
- Epochs

**Comment by firasdib** [comment]
For anyone looking for a workaround meanwhile, this is what I've been using:

```java
            final MultiLayerUpdater updater = (MultiLayerUpdater) model.getUpdater();

            for (final UpdaterBlock updaterBlock : updater.getUpdaterBlocks()) {
                final Adam config = (Adam) updaterBlock.getGradientUpdater()
                        .getConfig();

                // Iteration count is unused, any value works.
                config.applySchedules(-1, config.getLearningRate() * lrDecayFactor);
            }
```

**Comment by stolsvik** [comment]
Definitely agree on the log output of learning rate. Both on the schedule, and on any other variant that changes the learning rate. E.g. the parameters lrPolicyDecayRate(..) along with lrPolicyPower(..) aren't immediately obvious, and I would personally have a hard time understanding how the learning rate actually changed over time.

**Comment by stolsvik** [comment]
Since epochs aren't really a part of the learning (you only get that implicit by the fit(iterator)), a method whereby the user could notify the model that an epoch had started could maybe be of use. E.g. model.epochStart().

This could also be of use for things like ScoreIterationListener, as it would then know the difference between a mini batch, and an epoch.

The model should keep a counter of the epochs, and pass that to any listeners. Note that when epoch 0 was started, this is before *any* learning had taken place. I personally find it nice to know how the score and accuracy etc are right after weight inits etc.

This method would obviously be invoked automatically when using the fit(iterator), while if you used fit(dataset), you'd have to signal to the model yourself when whatever you consider an epoch starts.

**Comment by AlexDBlack** [comment]
This is now available as of 1.0.0-alpha
